tests/test_sbr_calculator.py

```diff
@@ -0,0 +1,524 @@
+    user_inputs = request.getfixturevalue(user_inputs)
+    sbr_val, _, _ = get_sbr_scores(user_inputs)
+    assert sbr_val == expected_sbr_score
+
+
+@pytest.mark.parametrize(
+    "user_inputs, expected_sbr",
+    [
+        ("test_smartest_home", ("A")),
+        ("test_sbr_b", ("B")),
+        ("test_solar_ev_battery_hp", ("C")),
+        ("test_solar_battery_power_diverter", ("C")),
+        ("test_hp_secondary_heating_hot_water", ("D")),
+        ("test_smart_meter_only", ("G")),
+        ("test_no_smart_meter", ("G")),
+    ],
+)
+def test_sbr_calculator(user_inputs, expected_sbr, request):
+    user_inputs = request.getfixturevalue(user_inputs)
+    _, sbr, _ = get_sbr_scores(user_inputs)
+    assert sbr == expected_sbr
+
+
+@pytest.mark.parametrize(
+    "user_inputs, expected_archetype",
+    [
+        ("test_smartest_home", (FlexArchetype.GOLD_FLEXER)),
+        ("test_sbr_b", (FlexArchetype.GOLD_FLEXER)),
+        ("test_solar_ev_battery_hp", (FlexArchetype.GOLD_FLEXER)),
+        ("test_solar_battery_power_diverter", (FlexArchetype.STRONG_FLEXER)),
+        ("test_hp_secondary_heating_hot_water", (FlexArchetype.GOOD_FLEXER)),
+        ("test_smart_meter_only", (FlexArchetype.LOW_TECH_FLEXER)),
+        ("test_no_smart_meter", (FlexArchetype.UNTAPPED_FLEXER)),
+    ],
+)
+def test_archetype_calculator(user_inputs, expected_archetype, request):
+    user_inputs = request.getfixturevalue(user_inputs)
+    _, _, archetype = get_sbr_scores(user_inputs)
+    assert archetype == expected_archetype
+
+
+def test_all_combinations():
+    """Test valid outputs for all combinations of get_sbr_scores inputs"""
+
+    input_combinations = itertools.product(
+        [True, False],  # smart_meter
+        [True, False],  # smart_ev_charger
+        EVChargerPower.__members__.values(),
+        [True, False],  # smart_v2g_enabled
+        [True, False],  # home_battery
+        BatterySize.__members__.values(),
+        [True, False],  # solar_pv
+        SolarInverterSize.__members__.values(),
+        [True, False],  # electric_heating
+        HeatingSource.__members__.values(),
+        HotWaterSource.__members__.values(),
+        [True, False],  # secondary_heating
+        [True, False],  # secondary_hot_water
+        [True, False],  # integrated_control_sys
+    )
+
+    for inputs in input_combinations:
+        user_inputs = UserInputs(*inputs)
+
+        if (
+            (not inputs[1] and inputs[2] != EVChargerPower.NONE)
+            or (not inputs[4] and inputs[5] != BatterySize.NONE)
+            or (not inputs[6] and inputs[7] != SolarInverterSize.NONE)
+            or (inputs[1] and inputs[2] == EVChargerPower.NONE)
+            or (inputs[4] and inputs[5] == BatterySize.NONE)
+            or (inputs[6] and inputs[7] == SolarInverterSize.NONE)
+        ):
+            with pytest.raises(AssertionError):
+                sbr_normalised, sbr, flex_archetype = sbr_score(*inputs)
+
+        else:
```

Comment:
A better way to write this test is to wrap this up as a pytest Test class:

class TestAllCombination:
    @pytest.mark.xfail(raises=AssertionError)
    def test_raises_exceptions_for_bad_inputs(self):
        """run logic that will fail test"""
    def test_assert_expected_types(self):
        assert isinstance(x, float), "x is not float"
        assert isinstance(x, str), "x is not str"

    def test_assert_value_in_list(self):
         assert x in y, "x must be in y"
         etc.

This breaks down the test into individual tests, each tests doing 1 thing, so that when things break you can easily find trace back to the source where bugs occur and fixes.
<img width="960" alt="IDE" src="https://github.com/centrefornetzero/smart-building-rating-calculator/assets/12820124/dc49a763-b039-40dd-ad4a-5d87e5536d35">

Doing that also makes unit testing work with many IDEs test plugins to quickly navigate failing tests and find out why tests are failing.

The purpose of unit tests is not just for one to check logic, is also to guard against regression tests, i.e. if a new developer comes in and make changes and accidentally break codes, it helps them to quickly find out where things are breaking and make fixes.

Reply:
Hopefully better now!